code/modules/research/weaponsanalyzer.dm

```diff
@@ -0,0 +1,282 @@
+	icon = 'icons/obj/machines/research.dmi'
+	icon_state = "weapon_analyzer"
+	density = 1
+	anchored = 1
+	use_power = 1
+	idle_power_usage = 60
+	active_power_usage = 2000
+
+	var/obj/item/gun/gun = null
+	var/obj/item/item = null
+	var/obj/item/device/laser_assembly/assembly = null
+	var/process = FALSE
+
+	component_types = list(
+			/obj/item/stock_parts/scanning_module = 2,
+			/obj/item/stock_parts/capacitor = 1,
+			/obj/item/stock_parts/console_screen = 1
+		)
+
+/obj/machinery/weapons_analyzer/examine(mob/user)
+	..()
+	var/name_of_thing
+	if(gun)
+		name_of_thing = gun.name
+	else if(item)
+		name_of_thing = item.name
+	to_chat(user, span("notice", "It has [name_of_thing ? "[name_of_thing]" : "nothing"] attached."))
+
+/obj/machinery/weapons_analyzer/attackby(var/obj/item/I, var/mob/user as mob)
+	if(!I || !user || !ishuman(user))
+		return
+
+	var/mob/living/carbon/human/H = user
+
+	if(istype(I, /obj/item/gun))
+
+		if(!check_gun(user))
+			return
+
+		gun = I
+
+		H.drop_from_inventory(I)
+		I.forceMove(src)
+		update_icon()
+	else if(istype(I, /obj/item/device/laser_assembly))
+		if(!check_gun(user))
+			return
+
+		var/obj/item/device/laser_assembly/A = I
+		A.ready_to_craft = TRUE
+		assembly = A
+		H.drop_from_inventory(I)
+		I.forceMove(src)
+		A.analyzer = WEAKREF(src)
+		update_icon()
+	else if(istype(I, /obj/item/laser_components))
+		if(!assembly)
+			to_chat(user, span("warning", "\The [src] does not have any assembly installed!"))
+			return
+		if(process)
+			to_chat(user, span("warning", "\The [src] is busy installing component!"))
+			return
+		var/success = assembly.attackby(I, user)
+		if(!success)
+			return
+		message_admins("success: [success]")
+		if(success == 2)
+			playsound(loc, 'sound/machines/weapons_analyzer_finish.ogg', 75, 1)
+			addtimer(CALLBACK(src, .proc/reset), 32)
+		else
+			playsound(loc, 'sound/machines/weapons_analyzer.ogg', 75, 1)
+			addtimer(CALLBACK(src, .proc/reset), 15)
+		process = TRUE
+		update_icon()
+	else if(I)
+		item = I
+		H.drop_from_inventory(I)
+		I.forceMove(src)
+		update_icon()
+
+/obj/machinery/weapons_analyzer/attack_hand(mob/user as mob)
+	user.set_machine(src)
+	ui_interact(user)
+
+/obj/machinery/weapons_analyzer/proc/reset()
+	process = FALSE
+	update_icon()
+
+/obj/machinery/weapons_analyzer/proc/check_gun(var/mob/user)
+	if(gun)
+		to_chat(user, span("warning", "\The [src] already has \the [gun] mounted. Remove it first."))
+		return FALSE
+	if(assembly)
+		to_chat(user, span("warning", "\The [src] already has \the [assembly] mounted. Remove it first."))
+		return FALSE
+	if(item)
+		to_chat(user, span("warning", "\The [src] already has \the [item] mounted. Remove it first."))
+		return FALSE
+
+	return TRUE
+
+/obj/machinery/weapons_analyzer/verb/eject()
+	set name = "Eject Inserted Item"
+	set category = "Object"
+	set src in view(1)
+
+	if(!usr || usr.stat || usr.lying || usr.restrained() || !Adjacent(usr))
+		return
+	
+	if(gun)
+		gun.forceMove(usr.loc)
```

Comment:
should probably use get_turf(usr) here, or try putting it in their hands

Reply:
or `get_turf(this)` if you want to behave like some other machines.

Reply:
That would drop it under the machine, and then you won't see it

Reply:
Pretty sure `/obj/machinery` by default is on a layer below `/obj/item`